File: javascript/packages/linter/src/rules/html-no-unknown-tag.ts
import { ParserRule } from "../types.js"
import { BaseRuleVisitor, isKnownHTMLElement, isKnownSVGElement, isKnownMathMLElement, isCustomElement } from "./rule-utils.js"
import { getTagLocalName, isHTMLOpenTagNode } from "@herb-tools/core"

import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, ParseResult, ParserOptions } from "@herb-tools/core"

const FOREIGN_CONTENT_TAGS = new Set(["svg", "math"])

function isComponentElement(node: HTMLElementNode): boolean {
  if (!isHTMLOpenTagNode(node.open_tag)) return false

  const rawTagName = node.open_tag.tag_name?.value
  if (!rawTagName) return false

  return /^[A-Z]/.test(rawTagName)
}

class NoUnknownTagVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (!tagName) {
      super.visitHTMLElementNode(node)
      return
    }

    if (FOREIGN_CONTENT_TAGS.has(tagName)) {
      return
    }

    if (isComponentElement(node)) {
      super.visitHTMLElementNode(node)
      return
    }

    if (!isCustomElement(tagName) && !isKnownHTMLElement(tagName) && !isKnownSVGElement(tagName) && !isKnownMathMLElement(tagName) && node.open_tag?.tag_name) {
      let message = `Unknown HTML tag \`<${tagName}>\`. This is not a standard HTML element.`

      if (tagName.includes("_")) {
        const suggestedName = tagName.replace(/_/g, "-")
        message += ` Did you mean \`<${suggestedName}>\`? Custom elements must contain a hyphen.`
      }

      this.addOffense(
        message,
        node.open_tag.tag_name.location,
      )
    }

    super.visitHTMLElementNode(node)
  }
}

export class HTMLNoUnknownTagRule extends ParserRule {
  static ruleName = "html-no-unknown-tag"
  static introducedIn = this.version("0.9.3")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: true,
      severity: "warning",
      exclude: ["**/*.xml.erb"],
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return {
      action_view_helpers: true,
      dot_notation_tags: true
    }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoUnknownTagVisitor(this.ruleName, context)

    visitor.visit(result.value)

    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/html_no_unknown_tag.rs
use crate::utils::html_data::{is_custom_element, is_known_html_element, is_known_mathml_element, is_known_svg_element};
use crate::utils::tag_utils::{get_open_tag, get_open_tag_name_token, get_tag_local_name};

use herb::nodes::HTMLElementNode;
use herb::Visitor;

const FOREIGN_CONTENT_TAGS: &[&str] = &["svg", "math"];

rule_visitor!(NoUnknownTagVisitor);
define_parser_rule!(
  HTMLNoUnknownTagRule,
  "html-no-unknown-tag",
  Warning,
  NoUnknownTagVisitor,
  exclude: ["**/*.xml.erb"],
  parser_options: { action_view_helpers: true, dot_notation_tags: true },
  introduced_in: "0.9.3"
);

fn is_component_element(node: &HTMLElementNode) -> bool {
  get_open_tag(node)
    .and_then(|open_tag| open_tag.tag_name.as_ref())
    .map(|token| token.value.starts_with(|character: char| character.is_ascii_uppercase()))
    .unwrap_or(false)
}

impl Visitor for NoUnknownTagVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    let tag_name = match get_tag_local_name(node) {
      Some(tag_name) => tag_name,
      None => {
        self.walk_html_element_node(node);
        return;
      }
    };

    if FOREIGN_CONTENT_TAGS.contains(&tag_name.as_str()) {
      return;
    }

    if is_component_element(node) {
      self.walk_html_element_node(node);
      return;
    }

    let is_unknown =
      !is_custom_element(&tag_name) && !is_known_html_element(&tag_name) && !is_known_svg_element(&tag_name) && !is_known_mathml_element(&tag_name);

    if is_unknown {
      if let Some(tag_name_token) = get_open_tag_name_token(node) {
        let mut message = format!("Unknown HTML tag `<{}>`. This is not a standard HTML element.", tag_name);

        if tag_name.contains('_') {
          message.push_str(&format!(
            " Did you mean `<{}>`? Custom elements must contain a hyphen.",
            tag_name.replace('_', "-")
          ));
        }

        self.add_offense(message, tag_name_token.location.clone());
      }
    }

    self.walk_html_element_node(node);
  }
}
